src/basler_synced_cameras_node.cpp

```diff
@@ -0,0 +1,293 @@
+#include <ros/console.h>
+#include <ros/ros.h>
+
+#include <pylon/PylonIncludes.h>
+#include <pylon/gige/BaslerGigEInstantCamera.h>
+
+#include <XmlRpcValue.h>
+
+#include "basler_parameters.h"
+#include "image_publisher.h"
+
+std::vector<Pylon::CBaslerGigEInstantCamera> loadCameras(const XmlRpc::XmlRpcValue& cameras_yaml)
+{
+    Pylon::CTlFactory& tl_factory = Pylon::CTlFactory::GetInstance();
+    Pylon::DeviceInfoList_t devices;
+    ros::Rate r(0.5);
+
+    std::vector<Pylon::CBaslerGigEInstantCamera> cameras(cameras_yaml.size());
+    for (size_t index = 0; index < cameras_yaml.size(); ++index)
+    {
+        ROS_ASSERT_MSG(cameras_yaml[index].getType() == XmlRpc::XmlRpcValue::TypeStruct, "Bad yaml for camera %zu", index);
+        ROS_ASSERT_MSG(cameras_yaml[index].hasMember("serial"), "Camera %zu yaml missing serial", index);
+        std::string serial = std::to_string(int(cameras_yaml[index]["serial"]));
+
+        bool camera_found = false;
+        ros::Time timeout_time = ros::Time::now() + ros::Duration(5);
+        while(ros::ok() && !camera_found)
+        {
+            tl_factory.EnumerateDevices(devices);
+            for (size_t i = 0; i < devices.size(); ++i)
+            {
+                if (devices[i].GetSerialNumber().c_str() == serial)
+                {
+                    ROS_INFO_STREAM("Found camera " << index << " with matching serial " << serial);
+                    cameras[index].Attach(tl_factory.CreateDevice(devices[i]));
+                    camera_found = true;
+                    break;
+                }
+            }
+            if (!camera_found)
+            {
+                ROS_DEBUG_STREAM("Failed to find camera " << index << " with serial " << serial);
+                ROS_DEBUG_STREAM( devices.size() << " other serials found:");
+                for (size_t i = 0; i < devices.size(); ++i)
+                {
+                    ROS_DEBUG_STREAM(devices[i].GetSerialNumber());
+                }
+                if (ros::Time::now() > timeout_time)
```

Comment:
Do you think it makes more sense as an individual timeout for each camera or an overall timeout for connecting to all of the cameras?  Might be slightly cleaner, but this would obviously time out faster if you have a large number of sync'd cameras.

Reply:
In practice I think the effect is largely the same. When a camera is present, the time to connect is pretty fast, so the timeout only comes into effect for the first missing camera. I haven't actually seen cases where it takes some time to connect, but if that's a thing that happens, I don't think the time one camera takes to connect should affect the time allotted for another camera.